buildSrc/src/main/groovy/org/elasticsearch/gradle/test/DistroTestPlugin.java

```diff
@@ -0,0 +1,279 @@
+
+    // all distributions used by distro tests. this is temporary until tests are per distribution
+    private static final String PACKAGING_DISTRIBUTION = "packaging";
+    private static final String COPY_PACKAGING_TASK = "copyPackagingArchives";
+
+    private Version upgradeVersion;
+
+    @Override
+    public void apply(Project project) {
+        project.getRootProject().getPluginManager().apply(GlobalBuildInfoPlugin.class);
+        project.getPluginManager().apply(JdkDownloadPlugin.class);
+        project.getPluginManager().apply(DistributionDownloadPlugin.class);
+        project.getPluginManager().apply(VagrantBasePlugin.class);
+        
+        configureVM(project);
+
+        this.upgradeVersion = getUpgradeVersion(project);
+
+        // setup task to run inside VM
+        configureDistributions(project);
+        configureCopyPackagingTask(project);
+        configureDistroTest(project);
+        configureBatsTest(project, "oss");
+        configureBatsTest(project, "default");
+    }
+
+    private static Jdk createJdk(NamedDomainObjectContainer<Jdk> jdksContainer, String name, String version, String platform) {
+        Jdk jdk = jdksContainer.create(name);
+        jdk.setVersion(version);
+        jdk.setPlatform(platform);
+        return jdk;
+    }
+
+    private static Version getUpgradeVersion(Project project) {
+        String upgradeFromVersionRaw = System.getProperty("tests.packaging.upgradeVersion");
+        if (upgradeFromVersionRaw != null) {
+            return Version.fromString(upgradeFromVersionRaw);
+        }
+
+        // was not passed in, so randomly choose one from bwc versions
+        ExtraPropertiesExtension extraProperties = project.getExtensions().getByType(ExtraPropertiesExtension.class);
+
+        if ((boolean) extraProperties.get("bwc_tests_enabled") == false) {
+            // Upgrade tests will go from current to current when the BWC tests are disabled to skip real BWC tests
+            return Version.fromString(project.getVersion().toString());
+        }
+
+        ExtraPropertiesExtension rootExtraProperties = project.getRootProject().getExtensions().getByType(ExtraPropertiesExtension.class);
+        String firstPartOfSeed = rootExtraProperties.get("testSeed").toString().split(":")[0];
+        final long seed = Long.parseUnsignedLong(firstPartOfSeed, 16);
+        BwcVersions bwcVersions = (BwcVersions) extraProperties.get("bwcVersions");
+        final List<Version> indexCompatVersions = bwcVersions.getIndexCompatible();
+        return indexCompatVersions.get(new Random(seed).nextInt(indexCompatVersions.size()));
+    }
+
+    private static void configureVM(Project project) {
+        String box = project.getName();
+
+        // setup jdks used by the distro tests, and by gradle executing
+        
+        NamedDomainObjectContainer<Jdk> jdksContainer = JdkDownloadPlugin.getContainer(project);
+        String platform = box.contains("windows") ? "windows" : "linux";
```

Comment:
We already do have it on the extension, but using it here is tricky. This would be set after applying the plugin, yet we need it here during configuration. So we would have to change Jdk to allow setting the platform with a Provider<String>, and all the fallout from that api wise (would we then make all the settings of jdk lazy settable?). Additionally, we have to set it *somewhere*, and I think this would be silly to do so explicitly in every platform when we know it based on the box name (which we are also getting from the project name). I don't feel strongly about this, but I would rather use something simple like this for now.

Reply:
Yeah, let's punt on this for now.